File: Algorithms/p230_Kth_Smallest_Element_in_a_BST_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int res;
        vector<TreeNode*> stack;
        while(1){
            if(root){
                stack.push_back(root);
                root = root->left;
            }else{
                if(--k==0){
                    res = (stack.back())->val;
                    break;
                };
                root = (stack.back())->right;
                stack.pop_back();
            }
        }
        return res;
    }
};

File: Algorithms/p230_Kth_Smallest_Element_in_a_BST/p230_Kth_Smallest_Element_in_a_BST_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        stack = []
        while True:
            if root:
                stack.append(root)
                root = root.left
            else:
                k -= 1
                if k==0:
                    res = stack[-1].val
                    break
                root = stack[-1].right
                stack.pop()
        return res
